parlai/agents/fid/fid.py

```diff
@@ -0,0 +1,251 @@
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""
+Leveraging Passage Retrieval with Generative Models for Open Domain Question Answering.
+
+See https://arxiv.org/abs/2007.01282
+"""
+import torch
+import torch.nn.functional as F
+from typing import Tuple, Union, Optional, List, Dict, Any
+
+from parlai.core.dict import DictionaryAgent
+from parlai.core.opt import Opt
+from parlai.agents.transformer.transformer import TransformerGeneratorModel
+
+from parlai.agents.rag.modules import RagModel, Document, T5RagModel
+from parlai.agents.rag.rag import RagAgent
+from parlai.agents.rag.model_types import RagToken, get_forced_decoder_inputs
+
+
+class Fid(RagToken):
+    """
+    FiD mimics RAG Token interface in many ways; we simply need to adjust the
+    decoder inputs to not repeat, as FiD attends over all encoder outputs
+    jointly.
+    """
+
+    def get_initial_forced_decoder_input(
+        self,
+        bsz: int,
+        inputs: torch.LongTensor,
+        n_docs: int,
+        start_idx: int,
+        end_idx: int,
+        input_turns_cnt: Optional[torch.LongTensor] = None,
+    ) -> torch.Tensor:
+        return get_forced_decoder_inputs(
+            inputs, bsz, start_idx, end_idx, self.generation_model
+        )
+
+    def get_initial_decoder_input(self, input: torch.LongTensor) -> torch.LongTensor:
+        return input
+
+    def get_next_decoder_input(
+        self,
+        prev_input: torch.LongTensor,
+        selection: torch.LongTensor,
+        incr_state_inds: torch.LongTensor,
+    ) -> torch.LongTensor:
+        prev_input = torch.index_select(prev_input, 0, incr_state_inds)  # type: ignore
+        decoder_input = torch.cat([prev_input, selection], dim=-1)
+        return decoder_input  # type: ignore
+
+
+class FidModel(RagModel):
+    """
+    The FiD Model is a simpler version of the RAG Model.
+
+    We override the encoder and decoder methods to join encoder outputs, and
+    decode normally, respectively.
+    """
+
+    def __init__(self, opt: Opt, dictionary: DictionaryAgent, retriever_shared=None):
+        super().__init__(opt, dictionary, retriever_shared=retriever_shared)
+        self.rag_model_interface = Fid(opt, dictionary[dictionary.null_token])
+        self.embedding_size = opt['embedding_size']
+
+    def reorder_encoder_states(
+        self,
+        encoder_states: Tuple[torch.Tensor, ...],
+        indices: Union[List[int], torch.LongTensor],
+    ) -> Tuple[torch.Tensor, torch.Tensor, List[List[Document]], torch.Tensor]:
+        """
+        Reorder the encoder states.
+
+        Override TGM.reorder_encoder_states to make sure we only pass enc, mask.
+
+        See ``TorchGeneratorModel.reorder_encoder_states`` for a description.
+        """
+        enc, mask, *_ = encoder_states
+        return TransformerGeneratorModel.reorder_encoder_states(
+            self, (enc, mask), indices
+        )
```

Comment:
This may be out of the scope of this PR. But I feel like we could eliminate having to override this if encoder_states were a dict or **kwargs instead of a tuple.

_In_ the scope of this PR, you could remove the need for this function by modifying `TransformerGeneratorModel.reorder_encoder_states` to do `enc, mask = encoder_states[:2]`.

Reply:
I actually find the forcing nice, as it causes people to realize they need to handle reordering states. kwargs could hide something that isn't being properly shuffled

Reply:
i should probably clarify this docstring, as we're overriding `RagModel.reorder_encoder_states`, not `TGM`